Use HTML meta charset attribute value, if present, when the Context-Type header does not specify it.

dantraMSFT opened the pull request:
Fix https://github.com/PowerShell/PowerShell/issues/3267

The change enables checking for the charset meta attribute when the Content-Type header does not define the charset. When found, the text content is re-encoded to the target charset.

A new property has been added to BasicHtmlWebResponseObject and HtmlWebResponseObject to support verifying the encoding. Attempting to do so from the returned content is non-deterministic.
See https://github.com/PowerShell/PowerShell-Docs/pull/1543 for example usage.

NOTE: Three tests are pending the resolution for issue https://github.com/PowerShell/PowerShell/issues/2867

dantraMSFT committed:
Use HTML meta charset attribute value, if present, when the Context-Type header does not specify it.

JamesWTruher commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -734,6 +760,84 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+    Context  "BasicHtmlWebResponseObject Encoding tests" {
+        It "Verifies Invoke-WebRequest detects charset meta value when the ContentType header does not define it." {
+            $output = '<html><head><meta charset="gb2312"></head></html>'
+            $expectedEncoding = [System.Text.Encoding]::GetEncoding('gb2312')
```
I'm worried about platforms where these encodings don't exist (linux/mac), which will cause the tests to fail. Do additional steps need to be taken to register new encodings, or does that happen elsewhere?

dantraMSFT replied:
fixed

SteveL-MSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs:
```diff
@@ -444,12 +431,65 @@ internal static string DecodeStream(Stream stream, Encoding encoding)
+            if (isDefaultEncoding) do
+            {
+                // check for a charset meta element to override the default.
+                string searchString = "<meta charset=\"";
```
I think a regular expression might be better here

dantraMSFT replied:
Fixed

dantraMSFT committed:
Switch to regex for searching for meta/charset attribute
Add negative test to validate handling of invalid/unsupported encodings in StreamHelper.DecodeStream

dantraMSFT commented:
Moving to regex for detecting charset. Will reopen when done.

dantraMSFT committed:
Update tests to use encodings that are available on Linux

dantraMSFT committed:
[Feature]
Revise regex expression to be less restrictive of the charset attribute value and use a named group for the value.

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/CoreCLR/HtmlWebResponseObject.CoreClr.cs:
```diff
@@ -49,8 +49,41 @@ internal HtmlWebResponseObject(HttpResponseMessage response, Stream contentStrea
+        /// <value>
+        /// The Encoding used to decode the Content; otherwise, a null reference if the content is not text.
+        /// </value>
+        public Encoding Encoding { get; private set; }
```
Done

dantraMSFT committed:
[Feature]
Updated regex to handle http-equiv charset declarations
Update tests to validate http-equiv charset declarations

dantraMSFT committed:
[Heading]
Strip trailing whitespace.

dantraMSFT commented:
@TravisEz13 and @JamesWTruher, please take a look again. I've completed the changes to use regex.

dantraMSFT commented on src/Microsoft.PowerShell.Commands.Utility/commands/utility/WebCmdlet/StreamHelper.cs:
```diff
@@ -444,12 +432,74 @@ internal static string DecodeStream(Stream stream, Encoding encoding)
+            if (isDefaultEncoding) do
+            {
+                // check for a charset attribute on the meta element to override the default.
+                Match match = s_metaexp.Match(content);
```
@TravisEz13 - According to my reading of the page, I'm doing exactly what it says...

"To prevent recompilation, you should instantiate a single Regex object that is accessible to all code that requires it, as shown in the following rewritten example."

Since s_metaexp is static/readonly, there is only one compilation. That appears to be the point of the of the caching; avoiding recompilation.  Am I missing something?

dantraMSFT replied:
My review of the corefx shows that constructing and reusing an instance (regex.Match) is equivalent to calling the static Match method.  The constructor manages cache lookups/updates. The static Match method constructs a new instance and calls through the instance Match method.
The relevant static method is:
public static Match Match(string input, string pattern, RegexOptions options, TimeSpan matchTimeout)
{
    return new Regex(pattern, options, matchTimeout, true).Match(input);
}

dantraMSFT commented on test/powershell/Modules/Microsoft.PowerShell.Utility/WebCmdlets.Tests.ps1:
```diff
@@ -805,6 +831,245 @@ Describe "Invoke-WebRequest tests" -Tags "Feature" {
+    #region charset encoding tests
+
+    Context  "BasicHtmlWebResponseObject Encoding tests" {
+        It "Verifies Invoke-WebRequest detects charset meta value when the ContentType header does not define it." {
```
I did add test cases covering variations of <meta charset...?> and <meta http-equiv...>, including various newline variations and unquoted attribute value.

dantraMSFT commented:
@TravisEz13: would you merge this.  Thanks.

chuanjiao10 commented:
Do not rush to close,Please help look, there seems to be a problem, i updated issues 3267.

SteveL-MSFT commented:
@chuanjiao10 since this is already merged, @dantraMSFT will submit new PR to address your issue

dantraMSFT commented:
@chuanjiao10, see https://github.com/PowerShell/PowerShell/pull/4692

chuanjiao10 commented:
good 3q
